numba/core/types/iterators.py

```diff
@@ -103,5 +103,6 @@ def __init__(self, array_type):
         elif nd == 1:
             yield_type = array_type.dtype
         else:
-            yield_type = array_type.copy(ndim=array_type.ndim - 1)
+            # iteration semantics leads to C order layout
+            yield_type = array_type.copy(ndim=array_type.ndim - 1, layout='C')
```

Comment:
setting layout to `'A'` is probably the safest fix.

Reply:
Done, also added the example as a test.